readthedocs/search/api/v3/backend.py

```diff
@@ -0,0 +1,225 @@
+     This limit is only applied for projects given explicitly,
+     not when we default to search all projects.
+    """
+
+    def __init__(
+        self,
+        *,
+        request,
+        query,
+        arguments_required=True,
+        default_all=False,
+        max_projects=100
+    ):
+        self.request = request
+        self.query = query
+        self.arguments_required = arguments_required
+        self.default_all = default_all
+        self.max_projects = max_projects
+
+    @cached_property
+    def projects(self):
+        """
+        Return all projects used in this search.
+
+        If empty, it will search all projects.
+
+        :returns: A list of tuples (project, version).
+        """
+        projects = islice(self._get_projects_to_search(), self.max_projects)
+        # Make sure we are using just one version per-project,
+        # searching multiple versions of the same projects isn't supported yet.
+        projects_dict = dict(projects)
+        return list(projects_dict.items())
+
+    def search(self, **kwargs):
+        """
+        Perform the search.
+
+        :param kwargs: All kwargs are passed to the `PageSearch` constructor.
+        """
+        if not self._has_arguments and self.arguments_required:
+            return None
+
+        projects = {project.slug: version.slug for project, version in self.projects}
+        # If the search is done without projects, ES will search on all projects.
+        # If we don't have projects and the user provided arguments,
+        # it means we don't have anything to search on (no results).
+        # Or if we don't have projects and we don't allow searching all,
+        # we also just return.
+        if not projects and (self._has_arguments or not self.default_all):
+            return None
+
+        search = PageSearch(
+            query=self.parser.query,
+            projects=projects,
+            **kwargs,
+        )
+        return search
+
+    def _get_projects_to_search(self):
+        if not self._has_arguments:
+            if self.arguments_required:
+                return None
+            yield from self._get_default_projects()
+            return None
+
+        for value in self.parser.arguments["project"]:
+            project, version = self._get_project_and_version(value)
+            if version and self._has_permission(self.request.user, version):
+                yield project, version
+
+        for value in self.parser.arguments["subprojects"]:
+            project, version = self._get_project_and_version(value)
+
+            # Add the project itself.
+            if version and self._has_permission(self.request.user, version):
+                yield project, version
+
+            # If the user didn't provide a version, version_slug will be `None`,
+            # and we add all subprojects with their default version,
+            # otherwise we will add all projects that match the given version.
```

Comment:
Only if they exist though, right?

Reply:
if we don't a find a version with that slug, we fallback to the default version of that subproject

https://github.com/readthedocs/readthedocs.org/blob/adec99886dafd5fccba498bcab707b3a5cb0b40a/readthedocs/search/api/v3/backend.py#L123-L130

This is similar to what we do now when adding results from subprojects automatically